File: packages/schemas/src/structureCard.ts
/**
 * StructureCard: a reusable creative structure distilled from one or more
 * analyzed videos. Stored in the knowledge base and retrieved during
 * storyboard generation.
 *
 * TS canonical schema. Pydantic mirror lives in
 * `apps/api/app/schemas/structure_card.py` (snake_case API surface).
 */

export type StructureCardId = string;

export interface EditingAtom {
  /** Short label, e.g. "hook", "reveal", "callout", "transition", "payoff". */
  kind: string;
  /** Approximate duration in seconds. */
  durationSeconds: number;
  /** Optional notes about pacing, transitions, or text overlays. */
  notes?: string;
}

export interface StructureCard {
  id: StructureCardId;
  /** Short, memorable label for the structure pattern. */
  patternName: string;
  /** 2-3 sentence description. */
  summary: string;
  /** Style of the opening hook. */
  hookType: string;
  /** High-level story arc. */
  narrativeFlow: string;
  /** Aesthetic, pacing, transitions, framing. */
  visualStyle: string;
  /** Ordered building blocks of the structure. */
  editingAtoms: EditingAtom[];
  /** Transferable rules for new topics. */
  reusableRules: string[];
  /** job_id of the source upload this card was distilled from. */
  sourceVideoJobId: string;
  /** Scene IDs from the source VideoAnalysis that informed this card. */
  sourceSegments: string[];
  /** ISO 8601 UTC timestamp. */
  createdAt: string;
}

File: apps/api/app/schemas/structure_card.py
"""StructureCard: a reusable creative structure distilled from a VideoAnalysis.

Stored in the knowledge base and consumed during storyboard generation.
The TS canonical mirror lives in `packages/schemas/src/structureCard.ts`.
"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class EditingAtom(BaseModel):
    """One reusable editing building block within a StructureCard."""

    model_config = ConfigDict(extra="ignore")

    kind: str = Field(
        description="Short label like 'hook', 'reveal', 'callout', 'transition'."
    )
    duration_seconds: float = Field(ge=0.0)
    notes: Optional[str] = Field(
        default=None,
        description="Optional pacing / overlay / transition hints.",
    )


class StructureCard(BaseModel):
    """A reusable creative structure transferable across topics."""

    model_config = ConfigDict(extra="ignore")

    id: str = Field(description="UUID for this card.")
    pattern_name: str = Field(
        description="Short memorable label, e.g. 'hook-then-reveal-payoff'."
    )
    summary: str = Field(description="2-3 sentence plain-language description.")
    hook_type: str = Field(description="Style of the opening hook.")
    narrative_flow: str = Field(description="High-level story arc.")
    visual_style: str = Field(
        description="Aesthetic, pacing, transitions, framing."
    )
    editing_atoms: List[EditingAtom] = Field(
        default_factory=list,
        description="Ordered editing atoms describing the reusable structure.",
    )
    reusable_rules: List[str] = Field(
        default_factory=list,
        description="Transferable rules / dos and don'ts.",
    )
    source_video_job_id: str = Field(
        description="job_id of the upload this card was distilled from."
    )
    source_segments: List[str] = Field(
        default_factory=list,
        description="Scene IDs from the source VideoAnalysis that informed this card.",
    )
    created_at: datetime = Field(description="UTC timestamp when the card was built.")
